### PR TITLE
Rack 3.0.0 release

### PR DESCRIPTION
**Pull request by MSP-Greg**
### Description

Rack 3.0.0 was released today, and breaks CI.

This PR makes the minimal changes needed to pass CI with Rack 3.0.0.

Currently, the Gemfile is `gem "rack", ">= 1.6.13"`.  Not sure if we should add an Actions workflow that runs a couple of jobs with Rack 2.2.x?

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: test/test_rack_server.rb [hunk 1]
```diff
@@ -20,21 +20,22 @@ def call(env)
         @app.call(env)
       rescue Exception => e
         @exception = e
+STDOUT.syswrite "\n#{e.message}\n"
```
Comment (dentarg) [inline test/test_rack_server.rb:23]: Missing indentation?

**Comment by dentarg** [comment]
>Not sure if we should add an Actions workflow that runs a couple of jobs with Rack 2.2.x?

We should I think

File: test/test_cli.rb [hunk 2]
```diff
@@ -49,7 +49,7 @@ def test_control_for_tcp
     cli = Puma::CLI.new ["-b", "tcp://127.0.0.1:0",
                          "--control-url", url,
                          "--control-token", "",
-                         "test/rackup/lobster.ru"], @log_writer, @events
+                         "test/rackup/hello.ru"], @log_writer, @events
```
Comment (dentarg) [inline test/test_cli.rb:52]: Why this change?

File: test/test_config.rb [hunk 3]
```diff
@@ -16,6 +17,7 @@ def test_default_max_threads
   end
 
   def test_app_from_rackup
+    skip_if :rack3
     conf = Puma::Configuration.new do |c|
       c.rackup "test/rackup/hello-bind.ru"
```
Comment (dentarg) [inline test/test_config.rb:22]: Can we use the rackup gem instead of skipping the test? https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md (On phone atm so haven't checked myself)

File: test/test_rack_server.rb [hunk 4]
```diff
@@ -20,21 +20,22 @@ def call(env)
...
         [ 500, {}, ["Error detected"] ]
       end
     end
   end
 
   class ServerLint < Rack::Lint
     def call(env)
-      check_env env
+      check_env env if Rack::RELEASE < '3'
```
Comment (dentarg) [inline test/test_rack_server.rb:31]: Didn't see removal of this method mentioned in the changelog for rack 3 https://github.com/rack/rack/blob/main/CHANGELOG.md

fyi @ioquatix

Comment (dentarg) [reply test/test_rack_server.rb:31]: Looks like it was renamed to `check_environment` in https://github.com/rack/rack/commit/eef251d3758741fec514356c117ff791a4c23347

File: test/test_rack_server.rb (hunk 1 above)
Comment (MSP-Greg) [reply test/test_rack_server.rb:23]: Sorry, that was my debug code.  I'll remove.

File: test/test_cli.rb (hunk 2 above)
Comment (MSP-Greg) [reply test/test_cli.rb:52]: I believe all the tests involve a control server, so the app isn't important?

File: test/test_rack_server.rb (hunk 4 above)
Comment (MSP-Greg) [reply test/test_rack_server.rb:31]: Actually, `Wrapper.new(@app, env).check_environment`.  Added.

File: test/test_config.rb (hunk 3 above)
Comment (MSP-Greg) [reply test/test_config.rb:22]: Loading a gem that has webrick as a dependency seems odd.  1,249 downloads since 03-Aug-22.

**Comment by MSP-Greg** [comment]
Added two jobs to a new Rack_v2 workflow, runs at the same time as the RuboCop workflow, so it shouldn't add much time to CI

File: test/test_config.rb (hunk 3 above)
Comment (dentarg) [reply test/test_config.rb:22]: Oh right, yeah never mind...

File: test/test_cli.rb (hunk 2 above)
Comment (dentarg) [reply test/test_cli.rb:52]: Just trying to understand the motivation behind making the change (to learn something)... has `Rack::Lobster` been removed?

**Comment by MSP-Greg** [comment]
Rebased with [Prefix all environment variables with PUMA_](https://github.com/puma/puma/commit/75853f2520a2ee9014cf1a3f627764aba21e1210)

File: test/test_cli.rb (hunk 2 above)
Comment (MSP-Greg) [reply test/test_cli.rb:52]: It's now part of rackup.  https://github.com/rack/rackup/blob/main/lib/rackup/lobster.rb

Sorry, when Rack v3.0.0 was released, I was in the middle of some Puma code, and wasn't too happy.

Maybe we should add a CI job that allows pre-releases and all dependencies are `>=`.  Give us better 'early warning'?

Comment (dentarg) [reply test/test_cli.rb:52]: Sure

**Comment by nateberkopec** [comment]
Thanks for the work Greg!

**Comment by ioquatix** [comment]
I've added this puma-head-rack-v3 to my conformance tests:

https://github.com/socketry/rack-conform/runs/8246190238?check_suite_focus=true

It failed because you may not be supporting Array headers. Can you check?

**Comment by dentarg** [comment]
Added that as an issue so we don't forget to follow up https://github.com/puma/puma/issues/2931
